crates/git-changeset/README.md

```diff
@@ -0,0 +1,177 @@
+# Git Changeset
+
+Please note that this is a first draft, and is a little bit of a brain dump of the things that I (Caleb) have been thinking about over the last few days. I expect to come back and re-visit most of this document in the following days as we start to put together an initial implementation, and talk to some of the kind folks over at [Mercurial](https://www.mercurial-scm.org/) about their implementation.
+
+Changeset aims to provide a way of referring to a change over time. In git, a commit is an immutable structure. "rewriting" a commit requires replacing the entire entry. That new commit does not give you any way of seeing what comes before it. Algorithms like [git-range-diff](https://git-scm.com/docs/git-range-diff) provide a best effort approach for determining what has changed, but there is an inherent imprecision, and greater cost associated with this heuristic based approach.
+
+Changeset introduces a new inter-commit graph that can be used to traverse different revisions of a commit. Changeset gives us the ability to have a pointer to the latest revision of a commit.
+
+This library aims to be an initial implementation of Changesets for the GitButler project, which could then be iterated on in collaboration with the kind folks from [JJ](https://github.com/martinvonz/jj) and [Sapling](https://github.com/facebook/sapling), in order to provide one standard interface which we can use. Eventually the goal is to have something like changesets introduced into git upstream, so all choices are chosen with the goal of being "git friendly".
+
+We've chosen to initially implement changesets in Rust as its easy to integrate into other rust projects like GitButler, JJ, and Sapling. Once we've stabilized an API, and are looking to upstream into git core we might look into re-implementing in C if that is desired.
+
+If any interesting parties has feedback or suggestions, I encourage you to reach out to me [caleb@gitbutler.com](mailto:caleb@gitbutler.com) or join our [Discord server](https://discord.com/invite/MmFkmaJ42D).
+
+## Requirements
+
+- A user should be able to find the latest revision of a commit for a given commit.
+- A user should be able to list the different revisions for a given changeset.
+- Changesets should be able to be worked on by two different users remotely, with ways of merging together two diverged versions of a particular changeset.
+- Changesets should have a unique identifier that tools can use to refer to them by.
+- Refer to an arbitrary commit by a changeset, without rewriting it.
+- Changesets need to be easily GCed.
```

Comment:
What does this mean?

Reply:
We need to have some way of clearing up old and unnecessary change sets so we don't end up flooding the ODB